src/libcore/array.rs

```diff
@@ -62,6 +64,111 @@ unsafe impl<T, A: Unsize<[T]>> FixedSizeArray<T> for A {
     }
 }
 
+/// An iterator that moves out of an array.
+#[derive(Debug)]
+pub struct IntoIter<T, A: FixedSizeArray<T>> {
+    // Invariants: index <= index_back <= array.len()
+    // Only values in array[index..index_back] are alive at any given time.
+    // Values from array[..index] and array[index_back..] are already moved/dropped.
+    array: Option<A>,
+    index: usize,
+    index_back: usize,
+    _marker: PhantomData<T>,
+}
+
+impl<T, A: FixedSizeArray<T>> Drop for IntoIter<T, A> {
+    #[inline]
+    fn drop(&mut self) {
+        // Drop values that are still alive.
+        if let Some(array) = self.array.as_mut() {
+            let slice = array.as_mut_slice();
+            for p in &mut slice[self.index..self.index_back] {
+                unsafe { ptr::drop_in_place(p); }
+            }
+        }
+
+        // Prevent the array as a whole from dropping.
+        unsafe { ptr::write(&mut self.array, None); }
```

Comment:
This actually has the same hazard you raised for `nth` -- if there's a panic in one of the `drop_in_place` calls, we won't get to clobber `self.array`.  And AFAICT a panic in a drop function will still try to drop the struct members, or even locals if I were to `take()` the array out early.

I think this will need drop-prevention to be an additional wrapper on the array, to be unwound on its own.